ArmPlatformPkg: Update PL011 Uart.

apop5 opened the pull request:
# Description

PL011UartRead would wait for the passed buffer to be filled before the function would return,
which can result in undesired wait times when passing larger buffers.
Modified the read function to fill the buffer or return when no more data was available.

SerialPortInitialize for PL011 Uart was modified to ensure that the serial buffer did not have
any data waiting in the Rx Fifo to match expectations for a 16550 device. 16550 devices have a Rx Fifo
clear bit in the Fifo control register, but the PL011 does not. Ensuring the Fifo is cleared
when serial port initialize is run will prevent any stale data from being interpereted
as valid.

See PL011 specification for more information:
https://documentation-service.arm.com/static/5e8e36c2fd977155116a90b5

- [ ] Breaking change?
- [ ] Impacts security?
- [ ] Includes tests?

## How This Was Tested
Used in physical platforms for 2+ years.

## Integration Instructions
No integration necessary.

ardbiesheuvel reviewed (changes requested):
Thanks for these fixes.

Please split this into two patches, and use meaningful titles for each

apop5 commented:
> Thanks for these fixes.
> 
> 
> 
> Please split this into two patches, and use meaningful titles for each

@ardbiesheuvel 

Thanks for reviewing so quickly.
I split the commits and updated commit messages.

ardbiesheuvel reviewed (commented):
You claim that the behavior of the PL011 driver deviates from other UART drivers in EDK2. However, the NS16550 driver (arguably the most widely used one), does the following:

```
  for (Result = 0; NumberOfBytes-- != 0; Result++, Buffer++) {
    //
    // Wait for the serial port to have some data.
    //
    while ((SerialPortReadRegister (SerialRegisterBase, R_UART_LSR) & B_UART_LSR_RXRDY) == 0) {
      if (PcdGetBool (PcdSerialUseHardwareFlowControl)) {
        //
        // Set RTS to let the peer send some data
        //
        SerialPortWriteRegister (SerialRegisterBase, R_UART_MCR, (UINT8)(Mcr | B_UART_MCR_RTS));
      }
    }

    if (PcdGetBool (PcdSerialUseHardwareFlowControl)) {
      //
      // Clear RTS to prevent peer from sending data
      //
      SerialPortWriteRegister (SerialRegisterBase, R_UART_MCR, Mcr);
    }

    //
    // Read byte from the receive buffer.
    //
    *Buffer = SerialPortReadRegister (SerialRegisterBase, R_UART_RXBUF);
  }
```

Also, the `SerialPortRead()` return value is documented as 

```
  @retval 0                NumberOfBytes is 0.
  @retval >0               The number of bytes read from the serial device.
                           If this value is less than NumberOfBytes, then the read operation failed.
```

which suggests that it is a read that block until the requested number of bytes has been read.

ardbiesheuvel commented on ArmPlatformPkg/Library/PL011SerialPortLib/PL011SerialPortLib.c:
```diff
@@ -34,22 +34,35 @@ SerialPortInitialize (
+             &StopBits
+             );
+
+  if (!EFI_ERROR (Status)) {
```
No success handling please. Make this

```
if (EFI_ERROR (Status)) {
  return Status;
}
 ... rest of the code ...
return EFI_SUCCESS;
```

apop5 replied:
Updated in latest.

apop5 commented:
> You claim that the behavior of the PL011 driver deviates from other UART drivers in EDK2. However, the NS16550 driver (arguably the most widely used one), does the following:
> 
> ```
>   for (Result = 0; NumberOfBytes-- != 0; Result++, Buffer++) {
>     //
>     // Wait for the serial port to have some data.
>     //
>     while ((SerialPortReadRegister (SerialRegisterBase, R_UART_LSR) & B_UART_LSR_RXRDY) == 0) {
>       if (PcdGetBool (PcdSerialUseHardwareFlowControl)) {
>         //
>         // Set RTS to let the peer send some data
>         //
>         SerialPortWriteRegister (SerialRegisterBase, R_UART_MCR, (UINT8)(Mcr | B_UART_MCR_RTS));
>       }
>     }
> 
>     if (PcdGetBool (PcdSerialUseHardwareFlowControl)) {
>       //
>       // Clear RTS to prevent peer from sending data
>       //
>       SerialPortWriteRegister (SerialRegisterBase, R_UART_MCR, Mcr);
>     }
> 
>     //
>     // Read byte from the receive buffer.
>     //
>     *Buffer = SerialPortReadRegister (SerialRegisterBase, R_UART_RXBUF);
>   }
> ```
> 
> Also, the `SerialPortRead()` return value is documented as
> 
> ```
>   @retval 0                NumberOfBytes is 0.
>   @retval >0               The number of bytes read from the serial device.
>                            If this value is less than NumberOfBytes, then the read operation failed.
> ```
> 
> which suggests that it is a read that block until the requested number of bytes has been read.

I dropped this change for now, and this PR will just be clearing of the rx buffer.

I will go back and look at the usecase that the other changes were attempting to address.

apop5 commented:
> > You claim that the behavior of the PL011 driver deviates from other UART drivers in EDK2. However, the NS16550 driver (arguably the most widely used one), does the following:
> 
> > 
> 
> > ```
> 
> >   for (Result = 0; NumberOfBytes-- != 0; Result++, Buffer++) {
> 
> >     //
> 
> >     // Wait for the serial port to have some data.
> 
> >     //
> 
> >     while ((SerialPortReadRegister (SerialRegisterBase, R_UART_LSR) & B_UART_LSR_RXRDY) == 0) {
> 
> >       if (PcdGetBool (PcdSerialUseHardwareFlowControl)) {
> 
> >         //
> 
> >         // Set RTS to let the peer send some data
> 
> >         //
> 
> >         SerialPortWriteRegister (SerialRegisterBase, R_UART_MCR, (UINT8)(Mcr | B_UART_MCR_RTS));
> 
> >       }
> 
> >     }
> 
> > 
> 
> >     if (PcdGetBool (PcdSerialUseHardwareFlowControl)) {
> 
> >       //
> 
> >       // Clear RTS to prevent peer from sending data
> 
> >       //
> 
> >       SerialPortWriteRegister (SerialRegisterBase, R_UART_MCR, Mcr);
> 
> >     }
> 
> > 
> 
> >     //
> 
> >     // Read byte from the receive buffer.
> 
> >     //
> 
> >     *Buffer = SerialPortReadRegister (SerialRegisterBase, R_UART_RXBUF);
> 
> >   }
> 
> > ```
> 
> > 
> 
> > Also, the `SerialPortRead()` return value is documented as
> 
> > 
> 
> > ```
> 
> >   @retval 0                NumberOfBytes is 0.
> 
> >   @retval >0               The number of bytes read from the serial device.
> 
> >                            If this value is less than NumberOfBytes, then the read operation failed.
> 
> > ```
> 
> > 
> 
> > which suggests that it is a read that block until the requested number of bytes has been read.
> 
> 
> 
> I dropped this change for now, and this PR will just be clearing of the rx buffer.
> 
> 
> 
> I will go back and look at the usecase that the other changes were attempting to address.
> 
> 

@ardbiesheuvel can you please review the reduced changes?

ardbiesheuvel commented on ArmPlatformPkg/Library/PL011SerialPortLib/PL011SerialPortLib.c:
```diff
@@ -34,22 +34,33 @@ SerialPortInitialize (
+  }
+
+  while (SerialPortRead (&Scratch, sizeof (Scratch)) != 0) {
+  }
```
So without the other change, this is now a blocking read of 1 byte. Shouldn't you be doing something like

```
while (SerialPortPoll ()) {
  SerialPortRead();
}
```

instead?

apop5 replied:
Yes, of course it should, otherwise it would hang there.

Thanks, its updated
